frontend/src/pages/notebookController/StartServerModal.tsx

```diff
@@ -15,93 +29,183 @@ type StartServerModalProps = {
+      if (startShown) {
+        window.addEventListener('beforeunload', onUnload);
+      }
+      return () => window.removeEventListener('beforeunload', onUnload);
+    }, [startShown, onUnload]);
+
+    React.useEffect(() => {
+      if (spawnInProgress) {
+        if (!notebookStatus) {
+          return;
+        }
+        if (notebookStatus.currentStatus === EventStatus.IN_PROGRESS) {
+          setSpawnPercentile(notebookStatus.percentile);
+        } else if (notebookStatus.currentStatus === EventStatus.ERROR) {
+          setSpawnInProgress(false);
           setSpawnStatus({
             status: AlertVariant.danger,
-            title: 'Failed to redirect',
-            reason:
-              'For unknown reasons the notebook server was unable to be redirected to. Please check your notebook status.',
+            title: notebookStatus.currentEventReason,
+            description: notebookStatus.currentEventDescription,
           });
         }
-      }, 6000);
-    }
-    return () => {
-      if (timer) {
-        clearTimeout(timer);
       }
+    }, [notebookStatus, spawnInProgress]);
+
+    const renderProgress = () => {
+      let variant;
+      switch (spawnStatus?.status) {
+        case AlertVariant.danger:
+          variant = ProgressVariant.danger;
+          break;
+        case AlertVariant.success:
+          variant = ProgressVariant.success;
+          break;
+        case AlertVariant.warning:
+          variant = ProgressVariant.warning;
+          break;
+        default:
+          variant = undefined;
+      }
+      return (
+        <Progress
+          id="progress-bar"
+          value={spawnPercentile}
+          title={
+            notebookStatus?.events.length
+              ? notebookStatus.currentEvent
+              : 'Waiting for server request to start...'
+          }
+          variant={variant}
+        />
+      );
     };
-  }, [isNotebookRunning, notebookLink]);
-
-  const loading = () => (
-    <>
-      <Spinner isSVG aria-label="Starting server modal spinner" />
-      <Button key="cancel" variant="secondary" onClick={onClose}>
-        Cancel
-      </Button>
-    </>
-  );
-
-  const running = () => (
-    <p className="odh-notebook-controller__start-server-modal-text">
-      Server ready at {notebookLink}
-    </p>
-  );
-
-  const renderStatus = () => {
-    if (!spawnStatus) {
-      return null;
-    }
+
+    const renderStatus = React.useCallback(() => {
+      if (!spawnStatus) {
+        return null;
+      }
+      return (
+        <Alert isInline variant={spawnStatus.status} title={spawnStatus.title}>
+          <p>{spawnStatus.description}</p>
+        </Alert>
+      );
+    }, [spawnStatus]);
+
+    const renderButtons = () =>
+      !isNotebookRunning ? (
+        <Button key="cancel" variant="secondary" onClick={onClose}>
+          {spawnFailed ? 'Close' : 'Cancel'}
+        </Button>
+      ) : null;
+
+    const renderLogs = React.useCallback(() => {
```

Comment:
Two things about all your render functions ... 

1. Don't `useCallback` when you're going to use the function immediately in the same context -- there is nothing saved here by NOT passing the reference of the function around -- `useCallback` is a way not to recompute function references; so if you don't pass the reference to a hook dependency or to another component, the reference remapping is immaterial the lifecycle of the function
2. This component is getting large -- can we not make each of these render functions a Component?